proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,230 @@
+  ::
+
+    let
+      name1 = _ :: Text
+      age1 = _ :: Int
+      name2 = _ :: Text
+      age2 = _ :: Int
+
+      textExample1 = name1 <> " (age: " <> T.pack (show age1) <> ") encountered " <> name2 <> " (age: " <> T.pack (show age2) <> ")"
+
+      textExample2 = T.pack $ T.unpack name1 <> " (age: " <> show age1 <> ") encountered " <> T.unpack name2 <> " (age: " <> show age2 <> ")"
+
+      textExample3 = T.unwords
+        [ name1
+        , "(age: " <> T.pack (show age1) <> ")"
+        , "encountered"
+        , name2
+        , "(age: " <> T.pack (show age2) <> ")"
+        ]
+
+* ``printf`` is partial and unsafe, which especially safety-conscious people might always stay away from anyway. Using a safer ``printf`` like ``formatting`` induces a third-party dependency, which is admittedly lightweight, but isn't as seamless of an integration as native string interpolation would be
+
+* Quasiquotes induces a dependency on Template Haskell, which a lot of people avoid out of principle. Most QuasiQuoters also add a dependency on ``haskell-src-exts`` to parse arbitrary Haskell expressions, which could technically be avoided by using something like ``ghc-meta`` (`repo <https://github.com/noughtmare/ghc-meta>`_, `GHC issue <https://gitlab.haskell.org/ghc/ghc/-/issues/20862>`_), but this isn't in wide use yet.
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+When a new ``StringInterpolation`` extension is enabled, the following syntax:
+::
+
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+will be syntax sugar for:
+::
+
+  interpolateRaw "foo "#
+  . doInterpolate (f a b)
+  . interpolateRaw " bar "#
+  . doInterpolate (g x)
+  . interpolateRaw " baz "#
+  . doInterpolate name
+  $ interpolateEmpty
+
+where:
+::
+
+  class Interpolate s where
+    interpolateRaw :: Addr# -> s -> s
+    default interpolateRaw :: (IsString s, Semigroup s) => Addr# -> s -> s
+    interpolateRaw s = (fromString (unpackCString# s) <>)
+
+    interpolateEmpty :: s
+    default interpolateEmpty :: Monoid s => s
+    interpolateEmpty = mempty
+
+  class Interpolate s => InterpolateValueRunner s a where
+    doInterpolate :: a -> s -> s
+
+  class Interpolate s => InterpolateValue flag s a where
+    {-# MINIMAL interpolate | interpolatePrec #-}
+
+    interpolate :: Proxy flag -> a -> s
+    interpolate proxy a = interpolatePrec proxy 0 a interpolateEmpty
+
+    interpolatePrec :: Proxy flag -> Int -> a -> s -> s
```

Comment:
What is this for? Generally there isn't any explanation of the design of this series of typeclasses.